models/multiphase/d2q9_ShanChen/Dynamics.c.Rt

```diff
@@ -0,0 +1,217 @@
+    return rho;
+}
+
+CudaDeviceFunction void Run() {
+    // This defines the dynamics that we run at each node in the domain.
+
+    rho = calcRho();
+
+    switch (NodeType & NODE_BOUNDARY) {
+    case NODE_Solid:
+    case NODE_Wall:
+        BounceBack();
+        break;
+    }
+    if (NodeType & NODE_MRT) 
+    {
+        CollisionBGK();
+    }
+
+    // rho = calcRho();
+    neighbour_type = neighbour_type(0,0);
+}
+
+CudaDeviceFunction void BounceBack() {
+    // Method to reverse distribution functions along the bounding nodes.
+    real_t uf;
+    uf = f[3];
+    f[3] = f[1];
+    f[1] = uf;
+    uf = f[4];
+    f[4] = f[2];
+    f[2] = uf;
+    uf = f[7];
+    f[7] = f[5];
+    f[5] = uf;
+    uf = f[8];
+    f[8] = f[6];
+    f[6] = uf;
+}
+
+CudaDeviceFunction void CollisionBGK() {
+    // Here we perform a single relaxation time collision operation.
+    // We save memory here by using a single dummy variable
+    real_t u_eq[2], f_temp[9];
+  
+    real_t d = getRho();
+
+    vector_t u_eq_vect = getUeq(d);
+    u_eq[0] = u_eq_vect.x; 
+    u_eq[1] = u_eq_vect.y;
+
+    f_temp[0] = f[0];
+    f_temp[1] = f[1];
+    f_temp[2] = f[2];
+    f_temp[3] = f[3];
+    f_temp[4] = f[4];
+    f_temp[5] = f[5];
+    f_temp[6] = f[6];
+    f_temp[7] = f[7];
+    f_temp[8] = f[8];
+    SetEquilibrium(d, u_eq); //stores equilibrium distribution in f[0]-f[8]
+    f[0] = f_temp[0] - omega*(f_temp[0]-f[0]);	
+    f[1] = f_temp[1] - omega*(f_temp[1]-f[1]);
+    f[2] = f_temp[2] - omega*(f_temp[2]-f[2]);
+    f[3] = f_temp[3] - omega*(f_temp[3]-f[3]);	
+    f[4] = f_temp[4] - omega*(f_temp[4]-f[4]);
+    f[5] = f_temp[5] - omega*(f_temp[5]-f[5]);
+    f[6] = f_temp[6] - omega*(f_temp[6]-f[6]);	
+    f[7] = f_temp[7] - omega*(f_temp[7]-f[7]);
+    f[8] = f_temp[8] - omega*(f_temp[8]-f[8]);
+}
+
+CudaDeviceFunction vector_t getU() {
+    // This function defines the macroscopic velocity at the current node.
+    real_t d = getRho();
+    vector_t u;
+
+    vector_t F_ff = getF_ff();
+    vector_t F_sf = getF_sf();
+
+    u.x = ( f[8]-f[7]-f[6]+f[5]-f[3]+f[1] )/d + (GravitationX + F_ff.x + F_sf.x )*0.5 ;
+    u.y = (-f[8]-f[7]+f[6]+f[5]-f[4]+f[2] )/d + (GravitationY + F_ff.y + F_sf.y )*0.5 ;
+    u.z = 0;
+    return u;
+}
+
+CudaDeviceFunction vector_t getUeq(real_t d) {
+    vector_t F_ff = getF_ff();
+    vector_t F_sf = getF_sf();
+    vector_t u_eq;
+
+    u_eq.x = ( f[8]-f[7]-f[6]+f[5]-f[3]+f[1] )/d;
+    u_eq.y = (-f[8]-f[7]+f[6]+f[5]-f[4]+f[2] )/d;
+
+    u_eq.x += (GravitationX + (F_ff.x + F_sf.x)/d) / omega;
+    u_eq.y += (GravitationY + (F_ff.y + F_sf.y)/d) / omega;
+    u_eq.z = 0;
+
+    return u_eq;
+}
+
+
+// ------------------------------ pseudo-potential related code --------------------------------------------- //
+
+CudaDeviceFunction void calcPsi() {
+    // Calculate psi at each point so that pseudopotential force can be found.
+    
+    real_t d = calcRho();
+    // real_t d = getRho(); // this one crashes
```

Comment:
I do load "density" during "PsiIteration" but getRho() crashes.

Reply:
I think you might misunderstood me at some point: 
 - each time you call getRho you sum up f's.
- during single stage local class variables (Dynamics.c is an inside of a class) could be used to store some reusable data 
- in S-C type models density/velocity often might be evaluated number of times, even during single stage. there is a room for some time-saving

Do not store density between stages - memory transfer is much more time consuming than sum(f[i])

Reply:
ok, I can agree that some simple functions like sum(f[i]) can be sometimes quicker than loading memory.
However, I still dont understand why calcRho() which sums up 'f' works while getRho() crashes.
AFAIK getRho() shall read the rho from previous stage... or maybe it reads it from previous iteration?

Reply:
Into which class Dynamics.c gets injected?